New Data Column Sidecar Storage Design, Data Columns as a First-Class Citizen & Unit Testing

nalepae opened the pull request:
### Genesis of this PR
This PR was originally initiated due to the failure of Prysm super nodes on `peerdas-devnet-5`.
The issue stemmed from a lack of [inodes](https://en.wikipedia.org/wiki/Inode) on the volume storing Prysm data.

Until now, Prysm stored data column sidecars in the same way as blob sidecars: one file per sidecar.
While this design worked well for storing blobs (as there are only a few per block), it caused issues when storing data columns.

For a retention period of 4096 epochs, with blobs in every block and 128 data columns per block (super-node), this resulted in over 16 million files just for storing data columns.

### New data column sidecar storage design
This PR introduces a new method for storing data column sidecars, significantly reducing inode usage.
A full description of the new storage design can be found [here](https://github.com/prysmaticlabs/prysm/blob/data-columns-storage/beacon-chain/db/filesystem/doc.go).

This design is not yet fully optimized, and at least one aspect could be improved:
- **Volume Usage Optimization**: Certain data, such as the beacon block header, are written multiple times within the same file. Implementing factorization could help reduce redundancy and improve efficiency.



### Data columns as a first-class citizen
Previously, many structures designed for data columns were adapted to fit blob-related structures.
While this approach worked well, it often felt like forcing a square peg into a round hole.

With the new data column storage design, the differences became too significant to continue this approach.
Some structures (e.g., caching) were rewritten to better align with the specific needs of data columns.

### Unit testing
Numerous unit tests have been added—not specifically for this PR, but for the files it modifies.
Some unit tests are still missing, but since this PR already introduces around 4,000 new lines of code and touches 75 files, I thought this was a [good place to stop](https://www.youtube.com/watch?v=C6W7_Aa8JUM&ab_channel=GoodPlaceToStop).

nalepae committed:
DB Filesystem: Move all data column related code to `data_columns.go`

Only code move.

nalepae committed:
Implement data columns storage

nalepae commented on consensus-types/blocks/rosidecar.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+// ROSidecar represents a read-only sidecar with its block root.
+type ROSidecar struct {
```
@kasey this it the new data structure I introduced to solve the "slice of interfaces" golang issue.

kasey commented on beacon-chain/blockchain/process_block.go:
```diff
@@ -693,37 +614,25 @@ func (s *Service) areDataColumnsAvailable(ctx context.Context, root [32]byte, si
-		return nil
-	}
-
 	// Subscribe to newsly data columns stored in the database.
```
Random typo spotted `newsly`

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/7fd169da5f436c1fe9fabe40a8396576b13d3157

kasey commented on beacon-chain/blockchain/process_block.go:
```diff
@@ -760,42 +670,44 @@ func (s *Service) areDataColumnsAvailable(ctx context.Context, root [32]byte, si
-				return nil
-			}
 		case <-ctx.Done():
 			var missingIndices interface{} = "all"
```
This technique is like gothic vampire; subtle mix of "elegant" and "cursed".

kasey commented on beacon-chain/blockchain/receive_block.go:
```diff
@@ -237,21 +238,27 @@ func (s *Service) handleDA(
 	avs das.AvailabilityStore,
 ) (time.Duration, error) {
 	daStartTime := time.Now()
```
Here's an idea to reduce clutter computing this return value:
```
func (s *Service) handleDA(ctx context.Context, block interfaces.SignedBeaconBlock, blockRoot [32]byte, avs das.AvailabilityStore) (time.Duration, error) {
	var elapsed time.Duration
	defer func(start time.Time) {
		elapsed = time.Since(start)
		dataAvailWaitedTime.Observe(float64(elapsed.Milliseconds()))
	}(time.Now())
	if avs != nil {
		if err := s.isDataAvailable(ctx, blockRoot, block); err != nil {
			return elapsed, errors.Wrap(err, "is data available")
		}
		return elapsed, nil
	}
	...
}
```
and use `elapsed` in all the return statements.

nalepae replied:
The issue with this solution is that `dataAvailWaitedTime` will be updated even if the error is nil.

Here is my proposed solution: https://github.com/prysmaticlabs/prysm/pull/15061/commits/fa813349835f4f13d3daf741b101f51369fbec9c

kasey commented on beacon-chain/blockchain/receive_block.go:
```diff
@@ -237,21 +238,27 @@ func (s *Service) handleDA(
 	}
+
+	nodeID := s.cfg.P2P.NodeID()
+	if err := avs.IsDataAvailable(ctx, nodeID, s.CurrentSlot(), rob); err != nil {
```
Could you pass `nodeID` to the constructor of the `NewLazilyPersistentStoreColumn` so that it isn't necessary to change the signature of `IsDataAvailable`?

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/5c76dc6c3a96c8f7a105af8a02a6ba84a5f7e4ee.

kasey commented on beacon-chain/db/filesystem/data_column_cache.go:
```diff
@@ -0,0 +1,219 @@
+		return false
+	}
+
+	for indice := range indices {
```
`indice` is odd spelling for a single index. `idx`?

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/5c76dc6c3a96c8f7a105af8a02a6ba84a5f7e4ee.

kasey commented on beacon-chain/db/filesystem/test_utils.go:
```diff
@@ -0,0 +1,83 @@
+	DataColumnsParamsByRoot map[[fieldparams.RootLength]byte][]DataColumnParams
+)
+
+func CreateTestVerifiedRoDataColumnSidecars(t *testing.T, dataColumnParamsByBlockRoot DataColumnsParamsByRoot) ([]blocks.RODataColumn, []blocks.VerifiedRODataColumn) {
```
For blobs I wound up moving this sort of functionality into the `verification` package to preserve the condition that other packages do not directly create the `VerifiedRO*` types. I see this requires a `*testing.T` so that does provide some good safety guarantees, but could this file move entirely into the verification package?

https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/verification/filesystem.go#L9
https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/verification/fake.go#L11

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/30ee3a72257cacd540448b03138605a899753cb6.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+
+// Store stores data column sidecars into the database and asynchronously performs pruning.
+// The returned chanel is closed when the pruning is complete.
+func (dcs *DataColumnStorage) Store(dataColumnSidecars []blocks.VerifiedRODataColumn) error {
```
nitpick: `Store` is fine but we do often use `store` as a noun for something that abstracts storage. `Save` would avoid that ambiguity and be consistent with the blob store.

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/799ed429ca63a61e9aa53b8c9de257efca3d9eb2.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	// Initialize the count of the saved SSZ encoded data column sidecar.
+	storedCount := 0
+
+	log.Error("AAAA")
```
AAAA!

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/6d16ed4a89504b2c7c067f5a15ea7bd7b65b049b.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+
+		// Overwrite the existing error only if it is nil, since the close error is less important.
+		if closeErr != nil && err == nil {
+			err = closeErr
```
Hesitate to even suggest this as error on close is rare, but maybe log it?
```
if closeErr == nil {
        return
}
if err != nil {
        log.WithError(err).WithField("path", filePath).Error("error closing file")
        return
}
err = closeErr
```

nalepae replied:
I was inspired by a master for this one:
https://github.com/prysmaticlabs/prysm/blob/6d16ed4a89504b2c7c067f5a15ea7bd7b65b049b/beacon-chain/db/filesystem/blob.go#L179-L185

I don't have any strong opinion on that. But I like the fact the close error is really handled as a real error.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	countWritten, err := file.Write(bytes)
+	if err != nil {
+		return errors.Wrap(err, "write")
+	}
```
IMO we'll start switch to fsync by default for blobs before peerdas, so in this branch I'd just add `file.Sync` right around here.

nalepae replied:
I don't see what adding `file.Sync` will bring here, since a file lines after we hit the `defer` statement containing the `file.Close` call. And closing the file dumps data into the disk right?

kasey replied:
`file.Sync` is `fsync` under the hood: https://linux.die.net/man/2/fsync

Calling close does not call fsync - data can still be in a cache waiting to be fully committed as written by the disk controller. When you call fsync, the process will block until the data is fully committed. Without calling fsync you may wind up with incomplete writes if there is a hard restart (which could happen for instance if the k8s pod is rescheduled or OOMKilled). This can result in corruption in the data file, extra important to avoid here where we have a header that needs to be consistent with the entries.

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/60db9a4d763d6267081e99d5659356ba4c011350.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	}
+
+	// Read the encoded file version.
+	var encodedFileVersion [versionSize]byte
```
Instead of sticking this in every file, what if we just add `v1` to the root directory name and infer versions from that?

nalepae replied:
Personally, I like the idea to have a "self-described" file.
For a retention period of 4096 epochs and blobs in all blocks, it only adds 128 kB of storage.

kasey replied:
It isn't much data so I don't feel strongly about it.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+const (
+	version                                    int = 0x01
+	versionSize                                    = 1           //bytes
+	maxSszEncodedDataColumnSidecarSize             = 536_870_912 // 2**(4*8) / 8 bytes
```
`endoded`

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	version                                    int = 0x01
+	versionSize                                    = 1           //bytes
+	maxSszEncodedDataColumnSidecarSize             = 536_870_912 // 2**(4*8) / 8 bytes
+	encodedSszEncodedDataColumnSidecarSizeSize     = 4           // bytes (size of the endoded size of the SSZ encoded data column sidecar)
```
`endoded`

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/5dfef04a6b81ecadc7d2ded43241745a97de4aaa.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+		return nil, ErrWrongVersion
+	}
+
+	// Read the SSZ encoded data column sidecar size.
```
I would explain this in slightly more detail, like 
```
// BlobColumnSidecar is a variable sized ssz object, but all columns for a block will be the same size.
// Saving the size of the sidecars for the block at bytes [1:5] allows us to represent offsets as multiples of this value, for compactness.
```

nalepae replied:
Actually it's quite explained in `doc.go`. But OK!

Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/7b9eef4a4561e21a12d941436b6bc18e703ef14e.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "read SSZ encoded data column sidecar size")
+	}
+	if countRead != encodedSszEncodedDataColumnSidecarSizeSize {
```
Why do we need this check?

kasey replied:
On second thought it would be useful to detect that the file was truncated.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	ErrWrongFileSize                        = errors.New("wrong file size")
+	ErrTooManyDataColumns                   = errors.New("too many data columns")
+	ErrWrongSszEncodedDataColumnSidecarSize = errors.New("wrong SSZ encoded data column sidecar size")
+	ErrDataColumnSidecarsFromDifferentSlots = errors.New("data column sidecars from different slots")
```
I think most or all of these variables can be un-exported.

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/c6e8c1a380068f2687e19a2e9c1e78265bc02463.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "read data column indices")
+	}
+	if countRead != mandatoryNumberOfColumns {
```
In all of these places where we allocate a certain size buffer, read it into it, and then check the bytes read, I think we could make things more concise using https://pkg.go.dev/io#ReadFull

Or even better, make a single buffer that is the right size to hold all the metadata and read it all at once, and just have a single error that indicates that the metadata section was truncated. Then you can slice and decode out of that single buffer. This avoids multiple syscalls.

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/aa84fbc597af4022e25d9ea104dca80c728849f2.

There is still `1+n` `file.Read`s:
- 1 (unique) for metadata, then
- 1 per column.

I did not find a way to, with only one syscall, read some parts of the file non contiguously.
We could of course read all the file at once then only select columns we want, but it would be a waste of read.
Also, we could do only one read for contiguous columns. (For example, if columns 4, 5, 6, 9 are requested, 1 read for columns 4 to 6 and one read for column 9. Let's say it's an optimisation for later ...)

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/22b3327a97afc4b1a171f569ce24520520b66a26.

kasey replied:
I only meant that we should get the metadata read done in one shot, not that we should read the column(s) data too. What you did is great!

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	}
+
+	// Get the size of the file.
+	fileSize := fileStat.Size()
```
I think you can avoid this and save the extra syscall (blocking your goroutine). You can compute the size from the other information in the metadata.

nalepae replied:
I can, but I see this as a check than the file is actually not corrupted, just after:

```go
	// Get the size of the file.
	fileSize := fileStat.Size()

	// Check the file size is correct.
	if uint32(fileSize-headerSize)%sszEncodedDataColumnSidecarSize != 0 {
		return nil, errWrongFileSize
	}
```

But I agree it's a quite weak check compared to the cost of `fileStat.Size()`. Let's ditch it.

Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/22b3327a97afc4b1a171f569ce24520520b66a26.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	}
+
+	// Retrieve the statistics of the file.
+	fileStat, err := file.Stat()
```
See below comment on `Size` - `Stat` can be expensive, like opening another file.

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/commit/22b3327a97afc4b1a171f569ce24520520b66a26.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+func (dcs *DataColumnStorage) saveFilesystem(root [fieldparams.RootLength]byte, epoch primitives.Epoch, dataColumnSidecars []blocks.VerifiedRODataColumn) error {
+	// Compute the file path.
+	filePath := filePath(root, epoch)
+	exists, err := afero.Exists(dcs.fs, filePath)
```
Wouldn't this check be racy? You could have multiple goroutines trying to call `saveDataColumnSidecarsNewFile`, and the way `os.Create` works, it will actually truncate the file when it is "created" the second time.

nalepae replied:
Nice catch, thanks.

Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/bc017ed4930a56769546f03c6e0b4b9d2d07978c.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+	fileMu.Lock()
+	defer fileMu.Unlock()
+
+	if exists {
```
I think it would be cleaner to refactor `saveDataColumnSidecarsExistingFile` / `saveDataColumnSidecarsNewFile` to a single func that checks for file existence internally. When that func runs, it will be under the `fileMu` mutex lock. Also I think that these 2 functions could be rewritten so that they would be nearly identical, to the point where it makes sense to just factor out the file descriptor setup and metadata reading (or initialization) and then have a single code path for both cases - probably with a new `metadata` type that manages the differences wrt creating vs updating the metadata section.

nalepae replied:
Actually these two functions are already under the `fileMu` mutex (locked by the caller function, `saveFilesystem`).

Before writing 2 different functions (new and existing), I tried to have only one managing both cases.
The factorization was not worth the decrease of readability. That's why I ended up with writing 2 different sub-functions. (Even if some code, but not so much, is common to the two functions.)

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+		indices := make([]uint64, 0, len(metadata.indices))
+		for index, i := range metadata.indices {
+			// Skip if the data column is not saved.
+			if i < limit {
```
I get it, if the high bit is set, the value will be > 128. But the name `limit` for this variable threw me off, it's a little confusing to read. Maybe something like `nonZeroOffset` would be clear?

nalepae replied:
Yes I struggled to find a good name. I was not happy with `limit` as well.

Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/584ef60ab0963bf85db19db637e131610ef9f4e6.

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+		}
+
+		// Compute the offset of the data column sidecar.
+		offset := headerSize + int64(metadata.indices[index]-limit)*int64(metadata.sszEncodedDataColumnSidecarSize)
```
While having the zero value mean unset is intuitive and is clearly safer, I keep thinking if all the unset fields had their high bit set when the metadata section is initialized, you never have to do conversions on the offsets, so it actually simplifies the code a little. eg if the value < 128 it would be set, >= 128, unset.

nalepae replied:
Does the loss of safety and intuitiveness worth the removal on one integer addition (`+128`) and one subtraction (`-128`) in the code? 

I'm not convinced.

kasey replied:
I'm not either! And to be clear it isn't a question of "efficiency" or something, more about remembering that when you interact with this byte value you need to add/subtract.

I just had an idea I like though, what about encapsulating it in a type like this?

```
type StorageIndex byte

func (i StorageIndex) IsSet() bool

func (i StorageIndex) Position() int

func (i StorageIndex) Set(pos int)

```

nalepae replied:
Done in https://github.com/prysmaticlabs/prysm/pull/15061/commits/feafec2b7ef836306c9fd4fa4b44f1be173cb210.

kasey replied:
nice!

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+
+		// Read the SSZ encoded data column sidecar.
+		sszEncodedDataColumnSidecar := make([]byte, metadata.sszEncodedDataColumnSidecarSize)
+		count, err := file.Read(sszEncodedDataColumnSidecar)
+		if err != nil {
+			return nil, errors.Wrap(err, "read SSZ encoded data column sidecar")
+		}
+		if uint32(count) != metadata.sszEncodedDataColumnSidecarSize {
+			return nil, ErrWrongBytesWritten
+		}
+
+		// Unmarshal the SSZ encoded data column sidecar.
+		dataColumnSidecar := &ethpb.DataColumnSidecar{}
+		err = dataColumnSidecar.UnmarshalSSZ(sszEncodedDataColumnSidecar)
+		if err != nil {
+			return nil, errors.Wrap(err, "unmarshal SSZ encoded data column sidecar")
+		}
+
+		// Create a RO data column.
+		roDataColumnSidecar, err := blocks.NewRODataColumn(dataColumnSidecar)
+		if err != nil {
+			return nil, errors.Wrap(err, "new read only data column")
+		}
+
+		// Create a verified RO data column.
+		verifiedRODataColumn := blocks.NewVerifiedRODataColumn(roDataColumnSidecar)
```
Since this section creates a `VerifiedRO` type I would prefer for it to be in the `verification` package. We want to be careful to not leak storage details into the other package, but I want to be strict about where we call the `NewVerifiedRO` types so we can enforce it with a static analyzer. I'm making a note to self to write that analyzer now. I think [ReadFull](https://pkg.go.dev/io#ReadFull) could be helpful here to avoid the detail leakage - you pass the buffer and the `Reader` to the func in `verification` and it returns the verified ro type.

nalepae replied:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/02a6c034a7a2715e53c0c0dabdb06e0b7fa2f658 (not using `ReadFull`).

kasey replied:
Awesome!

kasey commented on beacon-chain/db/filesystem/data_column.go:
```diff
@@ -0,0 +1,966 @@
+			continue
+		}
+
+		if period == highestPeriodToPrune {
```
At first I missed the `period < lowestPeriodToPrune` below. IMO move that case before the `==` case to flatten out this big if statement (since you'll know it is equal as it isn't gt or lt).

nalepae replied:
Done in https://github.com/prysmaticlabs/prysm/pull/15061/commits/ffc794182f60719c04fb0f74c6270fc4344a5c1a

kasey commented on beacon-chain/sync/data_columns.go:
```diff
@@ -150,12 +150,15 @@ func RequestDataColumnSidecarsByRoot(
+func SaveDataColumns(sidecars []blocks.RODataColumn, dataColumnStorage *filesystem.DataColumnStorage) error {
+	verifiedRODataColumns := make([]blocks.VerifiedRODataColumn, 0, len(sidecars))
+	for _, sidecar := range sidecars {
+		verifiedRODataColumn := blocks.NewVerifiedRODataColumn(sidecar)
```
This seems to be bypassing verification altogether?

nalepae replied:
The verification is actually done just before the `SaveDataColumns` is called (in `RequestDataColumnSidecarsByRoot`).

So, `RequestDataColumnSidecarsByRoot` should return `[]VerifiedRODataColumns`, not `[]RODataColumns`.
==> The verification is done (✅), but the semantic is not respected (❌) ...

kasey commented on beacon-chain/sync/data_columns_reconstruct.go:
```diff
@@ -109,35 +105,41 @@ func (s *Service) reconstructDataColumns(ctx context.Context, verifiedRODataColu
 		}
 
 		verifiedRoDataColumn := blocks.NewVerifiedRODataColumn(roDataColumn)
```
Another case of minting verified types without running a verification stack.

nalepae committed:
Kasey comment: Fix typo

nalepae committed:
Kasey comment: Fix clutter

nalepae committed:
Kasey comment: `IsDataAvailable`: Remove `nodeID`.

nalepae committed:
Kasey comment: indice ==> index

nalepae committed:
Kasey comment: Move `CreateTestVerifiedRoDataColumnSidecars` in `beacon-chain/verification/fake`.

nalepae committed:
`Store` ==> `Save`.

nalepae committed:
Kasey comment: AAAA!

nalepae committed:
Kasey comment: Fix typo.

nalepae committed:
Kasey comment: Add comment.

nalepae committed:
Kasey commnet: Stop exporting errors for nothing.

nalepae committed:
Kasey comment: Read all metadata at once.

nalepae committed:
Kasey comment: Compute file size instead of reading it from stats.

nalepae committed:
Kasey comment: Lock mutexes before checking if the file exists.

nalepae committed:
Kasey comment: `limit` ==> `nonZeroOffset`.

nalepae committed:
Kasey comment: `DataColumnStorage.Get`: Set verified into the `verification package`.

nalepae committed:
Kasey comment: `prune` - Flatten the `==` case.

nalepae committed:
Kasey comment: Implement and use `storageIndices`.

nalepae committed:
`DataColumnsAlignWithBlock`: Move into its own file.

nalepae committed:
`DataColumnSidecar`: Rename variables to stick with
https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/das-core.md#datacolumnsidecar

nalepae committed:
Kasey comment: Add `file.Sync`.

nalepae committed:
`DataColumnStorage.Get`: Remove useless cast.

nalepae committed:
(Internal) Kasey comment: Set automatically the count of saved data columns.

kasey commented:
great work!
